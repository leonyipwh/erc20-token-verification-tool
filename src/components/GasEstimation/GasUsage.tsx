import { useEffect, useState } from 'react';
import Box from '@mui/material/Box';
import Stack from '@mui/material/Stack';
import CountUp from 'react-countup';
import { mean, std, round } from 'mathjs';
import { SparklinesLine, Sparklines } from 'react-sparklines';
import { useTheme } from '@mui/material/styles';
import Divider from '@mui/material/Divider';
import Typography from '@mui/material/Typography';

interface IStat {
  label: string;
  value: number;
}

const GasUsage = ({ gasUsage }: { gasUsage: number[] }) => {
  const theme = useTheme();
  const [data, setData] = useState<IStat[]>([]);

  useEffect(() => {
    if (!gasUsage.length) return;
    setData([
      {
        label: 'Average Gas Usage',
        value: round(mean(gasUsage))
      },
      {
        label: 'Gas Usage Deviation',
        value: round(std(gasUsage))
      }
    ]);
  }, []);

  return (
    <>
      {!gasUsage.length ? (
        <>No transaction record available</>
      ) : (
        <>
          <Box>
            <Sparklines
              data={gasUsage}
              height={25}
              margin={5}
            >
              <SparklinesLine
                style={{
                  stroke: theme.palette.primary.light,
                  fill: 'none',
                  strokeWidth: 0.6
                }}
              />
            </Sparklines>
          </Box>
          <Stack
            direction={{ xs: 'column', sm: 'row' }}
            justifyContent="space-around"
            divider={
              <Divider
                orientation="vertical"
                flexItem
              />
            }
          >
            {data.map((item, index) => {
              return (
                <Box
                  key={index}
                  sx={{ textAlign: 'center' }}
                >
                  <Typography
                    variant="subtitle2"
                    gutterBottom
                    component="span"
                    color="primary.main"
                  >
                    {item.label}
                  </Typography>
                  <Box sx={{ textAlign: 'center' }}>
                    <Typography
                      variant="h6"
                      component="span"
                    >
                      <CountUp
                        end={item.value}
                        duration={0.5}
                      />
                    </Typography>
                  </Box>
                </Box>
              );
            })}
          </Stack>
        </>
      )}
    </>
  );
};

export default GasUsage;
