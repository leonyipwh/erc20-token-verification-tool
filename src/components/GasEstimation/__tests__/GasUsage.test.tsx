import GasUsage from '../GasUsage';
import { render } from '@testing-library/react';

describe('<GasUsage>', () => {
  it('should render', async () => {
    jest.useFakeTimers();
    const data = [1000, 2000, 3000, 4000];
    const { asFragment } = render(<GasUsage gasUsage={data} />);
    jest.runAllTimers();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display "No transaction record available"', async () => {
    const data: number[] = [];
    const { asFragment } = render(<GasUsage gasUsage={data} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
