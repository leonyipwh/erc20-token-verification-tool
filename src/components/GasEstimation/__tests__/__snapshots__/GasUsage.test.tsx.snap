// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GasUsage> should display "No transaction record available" 1`] = `
<DocumentFragment>
  No transaction record available
</DocumentFragment>
`;

exports[`<GasUsage> should render 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <svg
      preserveAspectRatio="none"
      viewBox="0 0 240 25"
    >
      <g>
        <circle
          cx="5"
          cy="20"
          r="2"
          style="stroke: #42a5f5; stroke-width: 0; fill-opacity: .1; fill: none; pointer-events: auto;"
        />
        <circle
          cx="81.66666666666667"
          cy="15"
          r="2"
          style="stroke: #42a5f5; stroke-width: 0; fill-opacity: .1; fill: none; pointer-events: auto;"
        />
        <circle
          cx="158.33333333333334"
          cy="10"
          r="2"
          style="stroke: #42a5f5; stroke-width: 0; fill-opacity: .1; fill: none; pointer-events: auto;"
        />
        <circle
          cx="235"
          cy="5"
          r="2"
          style="stroke: #42a5f5; stroke-width: 0; fill-opacity: .1; fill: none; pointer-events: auto;"
        />
        <polyline
          points="5 20 81.66666666666667 15 158.33333333333334 10 235 5 235 20 5 20 5 20"
          style="stroke: #42a5f5; stroke-width: 0; fill-opacity: .1; fill: none; pointer-events: auto;"
        />
        <polyline
          points="5 20 81.66666666666667 15 158.33333333333334 10 235 5"
          style="stroke: #42a5f5; stroke-width: 0.6; stroke-linejoin: round; stroke-linecap: round; fill: none;"
        />
      </g>
    </svg>
  </div>
  <div
    class="css-aesejs-MuiStack-root"
  >
    <div
      class="MuiBox-root css-xi606m"
    >
      <span
        class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-nx9l4d-MuiTypography-root"
      >
        Average Gas Usage
      </span>
      <div
        class="MuiBox-root css-xi606m"
      >
        <span
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          <span>
            2500
          </span>
        </span>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-xi606m"
    >
      <span
        class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-nx9l4d-MuiTypography-root"
      >
        Gas Usage Deviation
      </span>
      <div
        class="MuiBox-root css-xi606m"
      >
        <span
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          <span>
            1291
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
